sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -140,7 +141,9 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
     try {
       createdTempDir.foreach { path =>
         val fs = path.getFileSystem(hadoopConf)
-        if (fs.delete(path, true)) {
+        // Sometimes (e.g., when speculative task is enabled), temporary directories may be
+        // left uncleaned, confirmTempDirDeleted can confirm deleteOnExit.
```

Comment:
Do you mean even if we delete the temp dir here, some tasks may re-create it later?

Reply:
Yeap, the speculative tasks are maybe still running after Job committed.